controller/src/main/java/io/pravega/controller/store/client/StoreClientFactory.java

```diff
@@ -61,7 +62,7 @@ private static CuratorFramework createZKClient(ZKClientConfig zkClientConfig) {
                 .namespace(zkClientConfig.getNamespace())
                 .zookeeperFactory(new ZKClientFactory())
                 .retryPolicy(new ExponentialBackoffRetry(zkClientConfig.getInitialSleepInterval(),
-                        zkClientConfig.getMaxRetries()))
+                        zkClientConfig.getMaxRetries(), CURATOR_MAX_SLEEP_MS))
```

Comment:
Bounding the maximum amount of time of an iteration makes sense, but what I'm thinking is that once the session expires, it makes no sense to have another iteration. Would it be possible to simply abort the retry mechanism in the presence of a session timeout?

Reply:
There are different retry policies we can use here:

https://curator.apache.org/curator-client/index.html

There is no paticular reason for using the exponential backoff one if we are relying on a single session. We might want to change the policy to `RetryOneTime` or `RetryUntilElapsed`.

Reply:
actually the retry policy is not used only for session expiration but even in connection loss. 
So we dont necessarily want session expiration to affect our decision.. 

what i was talking about was incorporating session expiration in our retry policy decision making.. let me update this PR with those changes so that it is a better discussion. 
we can then decide how best to tackle this.

Reply:
For example, lets say we create a barrier called sessionExpiryFuture:
```
/**
* This future is linked with the zookeeper session expiration and is completed whenever the session expires. 
*/
CompletableFuture<Void> sessionExpiryFuture = new CompletableFuture<>();
```

Now we create a wrapper class `ExponentialBackoffWithSession` which takes a boolean supplier for deciding whether to retry or not. 
we will pass the can retry function as follows:
```        
Supplier<Boolean> canRetry = () -> !sessionExpiryFuture.isDone(); 
```
```
class ExponentialBackoffWithSession extends ExponentialBackoffRetry {
        private final Supplier<Boolean> canRetry;

        public ExponentialBackoffWithSession(int baseSleepTimeMs, int maxRetries, int maxSleepMs, Supplier<Boolean> canRetry) {
            super(baseSleepTimeMs, maxRetries, maxSleepMs);
            this.canRetry = canRetry;
        }

        @Override
        public boolean allowRetry(int retryCount, long elapsedTimeMs, RetrySleeper sleeper) {
            boolean b = canRetry.get() && super.allowRetry(retryCount, elapsedTimeMs, sleeper);
            return b;
        }
    }
```

we can then create curator client as follows:
```		
        ExponentialBackoffWithSession retryPolicy = new ExponentialBackoffWithSession(initialSleepInterval, maxRetries, maxSleepMs, canRetry);

        //Create and initialize the curator client framework.
        CuratorFramework zkClient = CuratorFrameworkFactory.builder()
                .connectString(connectionString)
                .namespace(namespace)
                .zookeeperFactory(new ZKClientFactory())
                .retryPolicy(retryPolicy)
                .sessionTimeoutMs(sessionTimeoutMs)
                .build();

This client takes the wrapper retry policy which will not retry if the session has expired. (ref: canRetry definition)
```

And we can register a watch on zk session for expiration:
```			
            zkClient.getZookeeperClient().getZooKeeper().register(event -> {
                if (event.getState() == Watcher.Event.KeeperState.Expired) {
                    sessionExpiryFuture.complete(null);                }
            });
```